SpokeStack/TFLiteWakewordRecognizer.swift

```diff
@@ -0,0 +1,478 @@
+                    var frameWindowArray: Array<Float32> = []
+                    while !self.frameWindow.isEmpty {
+                        let f = try self.frameWindow.read()
+                        frameWindowArray.append(f)
+                    }
+                    var stateArray: Array<Float32> = []
+                    for _ in 0..<self.stateWidth {
+                        let f = try self.encodeState.read()
+                        stateArray.append(f)
+                    }
+                    _ = try frameWindowArray
+                        .withUnsafeBytes(
+                            {try model.copy(Data($0), toInputAt: Tensors.encode.rawValue)})
+                    _ = try stateArray
+                        .withUnsafeBytes(
+                            {try model.copy(Data($0), toInputAt: Tensors.state.rawValue)})
+                    
+                    /// calculate
+                    try model.invoke()
+                    
+                    /// outputs
+                    let encodeOutput = try model.output(at: Tensors.encode.rawValue)
+                    let encodeResults = encodeOutput.data.withUnsafeBytes { (pointer: UnsafePointer<Float32>) -> [Float32] in
+                        Array<Float32>(UnsafeBufferPointer(start: pointer, count: encodeOutput.data.count / 4))}
+                    self.encodeWindow.rewind().seek(self.encodeWidth)
+                    for r in encodeResults {
+                        try self.encodeWindow.write(r)
+                        //encodeCollector.append(r)
+                    }
+                    let stateOutput = try model.output(at: Tensors.state.rawValue)
+                    let stateResults = stateOutput.data.withUnsafeBytes { (pointer: UnsafePointer<Float32>) -> [Float32] in
+                        Array<Float32>(UnsafeBufferPointer(start: pointer, count: stateOutput.data.count / 4))}
+                    for r in stateResults {
+                        try self.encodeState.write(r)
+                        //stateCollector.append(r)
+                    }
+                } else {
+                    throw WakewordModelError.encode("encode.lite was not initialized")
+                }
+            }
+        }
+    }
+    
+    private func detect() throws -> Bool {
+        if self.encodeWindow.isFull {
+            do {
+                if let model = self.detectModel {
+                    /// inputs
+                    var encodeWindowArray: Array<Float32> = []
+                    self.encodeWindow.rewind()
+                    while !self.encodeWindow.isEmpty {
+                        let f = try self.encodeWindow.read()
+                        encodeWindowArray.append(f)
+                    }
+                    _ = try encodeWindowArray
```

Comment:
It would probably be a good idea to create a simple swift abstraction around these tensor reads/writes to simplify the code (later). Those abstractions could also be used to avoid the temporary array allocations inside the frame loop.